README.md

```diff
@@ -15,17 +15,57 @@ Implement the tool `st` that does exact pattern matching using a suffix tree. Te
+The correctness has been verified using the very same tests used in the previous project, which were the following:
+
+- Short patterns and genomes (chains), with length going from 0 to 10 characters (selected randomly)
+- Random chains
+- Random chains, having a chance of repeating the previous character.
+- Random chains, having a chance of having a long sequence made up by a single character.
+- Random chains, where a character cannot appear consecutively two or more times.
+- Fibonacci chains, made up by two different characters being concatenated following a Fibonacci distribution.
+
+All of these genomes (except the short ones) have a length between 2,000 and 10,000 characters, whilst the patters length varies between 3 and 200. Moreover, for each pair of chains the genome is adapted, forcing to include the pattern a random number of times between 0 and 10 (if the randomness have made up more). The inclusion of this patterns has a high probability of having four extreme situations:
+
+- Pattern matching at the beginning of the genome.
+- Pattern matching at the end of the genoms.
+- Two pattern matchings next to each other.
+- Two pattern matchings overlapping one with each other.
+
+Moreover, since we had a working implementation of the exact pattern searching from the previous project, we have used them to double-check when finding errors while testing.
 
 ## Running time
 
-*Describe experiments that verifies that your implementation of `st` uses no more time than O(n) or O(n²) (depending on the algorithm) for constructing the suffix tree and no more than O(m) for searching for a given read in it. Remember to explain your choice of test data. What are “best” and “worst” case inputs?*
+### Naive Construction Algorithm
+The worst case inputs would be strings which suffixes are prefixes of any bigger suffix (the very worst case is a string of only the same character e.g. "aaaaaaaaa"), as this means the algorithm will need to search and compare each individual character of each suffix and when it creates a new node it will already have compared all characters of the suffix. This means the worst case is O(n^2) because it has to compare each character in each suffix to another character (except for the suffix that is the whole string).
+
+The best case input would be a string with no repeating characters, as this would mean it only has to compare the first character of the suffix to the first character of each child of the root before inserting the suffix in a new child of the root. This means it is O(a*n) for best case, where a is the size of the alphabet and n is the size of the string.
+
+To prove our implementation takes at most O(n^2) we benchmarked the runtime of the construction of suffix trees of strings of varying length consisting of only the same character as this is the worst case input.
+
+![](./figs/construction.jpg)
+
+Looking at the figure, the quadratic shape is apparent.
```

Comment:
Really? It looks $O(n^{2.1})$ to me!

(It also looks $O(n^3)$, 'cause that shape is also what you get for a cubic curve).

It is _bloody_ hard to tell these polynomials apart. But you know, 'cause we've talked about it many times in this class, that you get out of that problem by plotting the observed running time divided by the expected. If you do $n^2/n^2 = 1$ you get a horizontal line, and it is easy to see if something is a horizontal line (or at least close enough that it doesn't matter).

I believe that it is quadratic, but not because of your figure. Your figure could be a lot of things, and I have no way of detecting if it is $O(n^2)$.